### PR TITLE
session: move more session vars to stmt context for retrying

### PR DESCRIPTION
**Pull request by jackysp**
### What problem does this PR solve? 
Some session variables should be in statement context e.g. prev-affected rows and last insert id. If not the behavior will be incorrect when retrying.

### What is changed and how it works?
Move the session variables to statement context.

### Check List 

Tests 

 - Unit test (TODO)

Code changes

 - Has exported function/method change

Related changes

 - Need to cherry-pick to the release branch

PTAL @tiancaiamao @disksing

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8034)

### CODE REVIEW

**Comment by jackysp** [comment]
```sql
create table t (i int key);
insert into t values (1);
begin;
insert into t values (10);
update t set i = i + row_count();
-- before the following commit, run `update t set i = i - 1;` in another session.
commit;
```
we got 
```
mysql> select * from t;
+----+
| i  |
+----+
| -1 |
|  9 |
+----+
2 rows in set (0.00 sec)
```
I think it should be
```
mysql> select * from t;
+----+
| i  |
+----+
| 1  |
| 11 |
+----+
2 rows in set (0.00 sec)
```

**Comment by jackysp** [comment]
PTAL @disksing @tiancaiamao

**Comment by disksing** [comment]
Sorry I'm not really familiar with `session.go`. Maybe @coocood can help you.

File: session/session.go [hunk 1]
```diff
@@ -488,8 +487,6 @@ func (s *session) retry(ctx context.Context, maxCnt uint) error {
 			} else {
 				log.Warnf("con:%d schema_ver:%d retry_cnt:%d query_num:%d", connID, schemaVersion, retryCnt, i)
 			}
-			s.sessionVars.StmtCtx = sr.stmtCtx
-			s.sessionVars.StmtCtx.ResetForRetry()
```
Comment (zimulala) [inline session/session.go]: Cloud we remove the function of "ResetForRetry" in `StmtCtx`?

File: session/session_test.go [hunk 2]
```diff
@@ -1254,6 +1254,24 @@ func (s *testSessionSuite) TestDelete(c *C) {
...
+	tk1 := testkit.NewTestKitWithInit(c, s.store)
+
+	tk.MustExec("create table t (i int);")
+	tk.MustExec("insert into t values (1);")
+	tk.MustExec("begin;")
+	tk.MustExec("insert into t values (10);")
+	tk.MustExec("update t set i = i + row_count();")
+	tk.MustQuery("select * from t;").Check(testkit.Rows("2", "11"))
+
+	tk1.MustExec("update t set i = 0 where i = 1;")
```
[the remark this answers is not in the document]
Comment (jackysp) [reply session/session_test.go]: Yes.

**Comment by jackysp** [comment]
PTAL @zimulala

**Comment by jackysp** [comment]
PTAL @coocood if you are free.

**Comment by crazycs520** [comment]
I found a problem when test this PR.
first,
```sql
create table t (i int);
insert into t values (1);
```

then:

|      | session 1                         | Session 2                     |
| ---- | --------------------------------- | ----------------------------- |
| Sql1 | insert into t values (10);         |                               |
| Sql2 | update t set i = i + row_count(); | update t set i = 0 where i=1; |

look for sql 2 in seesion 1 and seesion 2:

Assume， Update is divided into 2 steps: 

1. Get txn startTS from PD
2. update data

The time series is as follows:

| Time | session 1               | session 2               |
| ---- | ----------------------- | ----------------------- |
| T0   | Get txn startTS from PD |                         |
| T1   |                         | Get txn startTS from PD |
| T2   |                         | update data             |
| T3   | update data             |                         |
| T4   | conflict , do retry     |                         |

Then , the result is:

```sql
mysql root@127.0.0.1:test> select * from t;
+----+
| i  |
+----+
| 2  |
| 12 |
+----+
```

I think result should be: 

```sql
mysql root@127.0.0.1:test> select * from t;
+----+
| i  |
+----+
| 1  |
| 11 |
+----+
```



Cause of this is when session1 do retry, the prevent StmtCtx is `update` not` insert`.

File: sessionctx/stmtctx/stmtctx.go [hunk 3]
```diff
@@ -71,6 +71,10 @@ type StatementContext struct {
 		histogramsNotLoad bool
 		execDetails       execdetails.ExecDetails
 	}
+	// PrevAffectedRows is the affected-rows value(DDL is 0, DML is the number of affected rows).
+	PrevAffectedRows int64
+	// params for prepared statements
+	PreparedParams []interface{}
```
Comment (crazycs520) [inline sessionctx/stmtctx/stmtctx.go]: why change []types.Datum to []interface{} ?

File: executor/executor.go [hunk 4]
```diff
@@ -1175,6 +1175,7 @@ func ResetContextOfStmt(ctx sessionctx.Context, s ast.StmtNode) (err error) {
 	sc.MemTracker = memory.NewTracker(s.Text(), vars.MemQuotaQuery)
 	sc.NowTs = time.Time{}
 	sc.SysTs = time.Time{}
+	sc.PreparedParams = []interface{}{}
```
Comment (crazycs520) [inline executor/executor.go]: Maybe put `PreparedParams` in `SessionVars` can avoid re-allocating memory before every execution.

Comment (tiancaiamao) [reply executor/executor.go]: Why not set it to `nil`

File: expression/builtin_other.go [hunk 5]
```diff
@@ -790,7 +790,7 @@ func (b *builtinGetParamStringSig) evalString(row chunk.Row) (string, bool, erro
 	if isNull || err != nil {
 		return "", isNull, errors.Trace(err)
 	}
-	v := sessionVars.PreparedParams[idx]
+	v := sessionVars.StmtCtx.PreparedParams[idx].(types.Datum)
```
Comment (tiancaiamao) [inline expression/builtin_other.go]: Is `PreparedParams` always types.Datum?
If so, could we declare `type PreparedParams []types.Datum` for it

File: sessionctx/variable/session_test.go [hunk 6]
```diff
@@ -44,4 +44,9 @@ func (*testSessionSuite) TestSession(c *C) {
 	// For last insert id
 	ctx.GetSessionVars().SetLastInsertID(1)
 	c.Assert(ctx.GetSessionVars().LastInsertID, Equals, uint64(1))
+
```
Comment (tiancaiamao) [inline sessionctx/variable/session_test.go]: Any SQL cases for the change ?

**Comment by jackysp** [comment]
PTAL @tiancaiamao

**Comment by tiancaiamao** [comment]
LGTM

**Comment by jackysp** [comment]
PTAL @lysu @zimulala

**Comment by jackysp** [comment]
PTAL @lysu @zimulala

**Comment by jackysp** [comment]
PTAL @lysu @zimulala

**Comment by jackysp** [comment]
PTAL @lysu @zimulala

**Comment by jackysp** [comment]
PTAL @lysu @zimulala

**Pull request merged**